wasmplugin/plugin.go

```diff
@@ -270,14 +270,28 @@ func (ctx *httpContext) OnHttpRequestBody(bodySize int, endOfStream bool) types.
 
 func (ctx *httpContext) OnHttpResponseHeaders(numHeaders int, endOfStream bool) types.Action {
 	defer logTime("OnHttpResponseHeaders", currentTime())
-
+	tx := ctx.tx
 	if ctx.interruptionHandled {
 		// Handling the interruption (see handleInterruption) generates a HttpResponse with the required status code.
 		// If handleInterruption is raised during OnHttpRequestHeaders or OnHttpRequestBody, the crafted response is sent
 		// downstream via the filter chain, therefore OnHttpResponseHeaders is called.
-		// We expect a response that is ending the stream, with exactly one header (:status) and no body.
+		// We expect a response that is ending the stream, with a :status header equal to the status defined in the interruption
+		// and no body.
 		// See https://github.com/corazawaf/coraza-proxy-wasm/pull/126
-		if numHeaders == 1 && endOfStream {
+		status, err := proxywasm.GetHttpResponseHeader(":status")
```

Comment:
I'm trying to understand this

https://github.com/corazawaf/coraza-proxy-wasm/pull/126#discussion_r1060550063

but not too sure, why isn't this just this?

```
if ctx.interruptionHandled {
  // Already sent local response in request stage so we need to return pause in response stage
  return types.ActionPause
}
```

Given how explicit the code is I feel there must be a good reason but don't understand from the comment

Reply:
When an interruption is raised, we generate a local response via `proxywasm.SendHttpResponse`. Depending on the phase in which the response is generated, this response may come back following the filter chain. So, the `OnHttpResponseHeaders` callback, can be triggered with our own deny response that is being sent to the client.
This is why inside `OnHttpResponseHeaders` we have an extra logic `if ctx.interruptionHandled`: if we have an interruption, and we have a response that is going back to the client with the status code of the interruption, we have to let it go. 
In any other case, we want to drop it, because it may be the upstream that is leaking data

Reply:
Is `OnHttpResponseHeaders` called twice, once with local response and once with upstream response?

Or otherwise, is it not deterministic whether `OnHttpResponseHeaders` is called with the local response or the upstream response? I'm trying to understand which is the case where the status is not the interruption.

For example if the condition is when interrupted in request headers phase, response headers is called with the direct response but when blocked in request body it is called with upstream response, we should be able to code that logic without checking status. But if it's some non-deterministic behavior then I guess we need to check it but that is surprising.

Reply:
To the best of what I understood back in the days (https://github.com/corazawaf/coraza-proxy-wasm/pull/126#discussion_r1060550063):
- If the interruption is raised during `OnHttpRequestHeaders/Body` we will have `OnHttpResponseHeaders` triggered by the local response that is going back.
- It does not happen when the interruption is raised during `OnHttpResponseHeaders`.

> I'm trying to understand which is the case where the status is not the interruption.

Sorry, I've not been clear enough about it. I thought the check was an extra precaution to be sure that we are not sending back the real response when an interruption is raised, but the `Interruption already handled, unexpected local response` scenario is something that should never be reached. It is just like these conditions that we have at the beginning of any callback:
```
if ctx.interruptionHandled {
   ctx.logger.Error().Msg("Interruption already handled")
   return types.ActionPause	
}
```

If we feel that this extra check is a useless overhead we can just allow any response generated by `OnHttpRequestHeaders/Body` that goes through `OnHttpResponseHeaders`.

Reply:
Cool - yeah I don't think of it as performance overhead really but it makes the code confusing. We're having this conversation in the first place because it's unclear why we have the check :) So if it's not needed, we shouldn't have it to prevent confusion and I'd stick with the normal simple one.

If it seems really useful to have a doublecheck, then the comment could all be rewritten to state that intent instead, e.g.

```
An interruption was raised in request processing, so this should be the local response we created then. We check the status once more to see if it is expected as a safety measure in case a bug or change in Envoy behavior causes a real response to instead be passed here.
```

Though I would probably still avoid it - reading that second sentence, it seems quite hand-wavy.

Reply:
I went for removing it, and looking when the interruption has been raised to just let the local response go. I only left a comment about this possible doublecheck.